### PR TITLE
Implement clustering algorithms: Bisecting KMeans, Gaussian Mixture, LDA

### LINKED ISSUE #45: [Clustering] Implement KMeans++

### LINKED ISSUE #46: [Clustering] Implement Gaussian Mixture

### LINKED ISSUE #48: [Clustering] Implement Latent Dirichlet Allocation (LDA)

### PR DESCRIPTION
**Pull request by menuka94**
Closes #45
Closes #46 
Closes #48

### CODE REVIEW

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java [hunk 1]
```diff
@@ -42,59 +56,102 @@ public ClusteringQueryHandler(ModelRequest request, StreamObserver<ModelResponse
...
+        String jsonString = jsonResults.collectAsList().toString();
 
-        featureDF = scalerModel.transform(featureDF);
-        featureDF = featureDF.drop("features");
-        featureDF = featureDF.withColumnRenamed("normalized_features", "features");
+        Gson gson = new Gson();
+        Type type = new TypeToken<List<ClusteringResult>>() {
+        }.getType();
+        List<ClusteringResult> results = gson.fromJson(jsonString, type);
+        log.info("results.size(): " + results.size());
```
Comment (hamersaw) [inline src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:137]: we need to reach some consensus on logging levels. i don't think this should be logged at the info level.

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java [hunk 2]
```diff
@@ -168,11 +335,29 @@ private void logRequest() {
...
+                log.info("\tMaxIterations: " + gaussianMixtureRequest.getMaxIterations());
+                break;
+            case LATENT_DIRICHLET_ALLOCATION:
+                LatentDirichletAllocationRequest latentDirichletAllocationRequest =
+                        request.getLatentDirichletAllocationRequest();
+                log.info("\tClusterCount: " + latentDirichletAllocationRequest.getClusterCount());
+                log.info("\tMaxIterations: " + latentDirichletAllocationRequest.getMaxIterations());
+                break;
+        }
         log.info("=======================================");
```
Comment (hamersaw) [inline src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:361]: i don't see the value in logging this. certainly not at the info level.

File: src/main/java/org/sustain/modeling/ClusteringQueryHandler.java [hunk 3]
```diff
@@ -130,6 +187,118 @@ private void buildModel() {
...
+            );
+        }
+    }
+
+    private Dataset<Row> preprocessAndGetFeatureDF() {
+        log.info("Preprocessing data");
+        ReadConfig readConfig = ReadConfig.create(sparkContext);
+        Dataset<Row> collection = MongoSpark.load(sparkContext, readConfig).toDF();
+        List<String> featuresList = new ArrayList<>(request.getCollections(0).getFeaturesList());
+        Seq<String> features = convertListToSeq(featuresList);
```
Comment (hamersaw) [inline src/main/java/org/sustain/modeling/ClusteringQueryHandler.java:271]: this is certainly a temporary solution

**Review by hamersaw — CHANGES_REQUESTED** [review]
there are logging issues. outside of that i can't comment on the correctness of the training process. i'm OK accepting it at face value.

File: src/main/java/org/sustain/handlers/ClusteringQueryHandler.java [hunk 4]
```diff
@@ -9,91 +9,152 @@
...
 public class ClusteringQueryHandler extends GrpcHandler<ModelRequest, ModelResponse> {
-
     private static final Logger log = LogManager.getFormatterLogger(ClusteringQueryHandler.class);
+    private final ModelRequest request;
+    private final StreamObserver<ModelResponse> responseObserver;
     private JavaSparkContext sparkContext;
 
     public ClusteringQueryHandler(ModelRequest request, StreamObserver<ModelResponse> responseObserver) {
         super(request, responseObserver);
+        this.request = request;
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/handlers/ClusteringQueryHandler.java:56]: Why are you reassigning this?
The call to `super(request, responseObserver)` assigns them to member variables of `GrpcHandler`, which you inherit.

File: src/main/java/org/sustain/handlers/ClusteringQueryHandler.java [hunk 5]
```diff
@@ -152,6 +325,44 @@ private void addClusterDependencyJars(JavaSparkContext sparkContext) {
         }
     }
 
+    private void logRequest(ModelType modelType) {
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/handlers/ClusteringQueryHandler.java:328]: Remove this function

**Review by inf0rmatiker — CHANGES_REQUESTED** [review]

File: src/main/java/org/sustain/handlers/ClusteringQueryHandler.java (hunk 4 above)
Comment (menuka94) [reply src/main/java/org/sustain/handlers/ClusteringQueryHandler.java:56]: Oh I've missed that.

File: src/main/resources/log4j.properties [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Root logger option
```
Comment (inf0rmatiker) [inline src/main/resources/log4j.properties:1]: nice

**Review by inf0rmatiker — APPROVED** [review]
We need to take a stab at a couple things this week:
- Providing an interface for model building
- Optimizing model building (reusing Spark Session, uploading deps in advance)
- Standardizing logging

Those should be delegated to a separate Issue though, so this branch is good to go.

**Pull request merged**